File: sol/solution/0301-0400/0301/Solution.erl
-spec remove_invalid_parentheses(S :: unicode:unicode_binary()) -> [unicode:unicode_binary()].
remove_invalid_parentheses(S) ->
    Chars = unicode:characters_to_list(S),
    {RmL, RmR} = count_removals(Chars, 0, 0),
    AnsSet = dfs(Chars, RmL, RmR, 0, [], sets:new()),
    sets:to_list(AnsSet).
-spec count_removals([integer()], integer(), integer()) -> {integer(), integer()}.
count_removals([], L, R) -> {L, R};
count_removals([$\( | T], L, R) -> count_removals(T, L+1, R);
count_removals([$) | T], L, R) when L > 0 -> count_removals(T, L-1, R);
count_removals([$) | T], L, R) -> count_removals(T, L, R+1);
count_removals([_ | T], L, R) -> count_removals(T, L, R).
-spec dfs([integer()], integer(), integer(), integer(), [integer()], sets:set()) -> sets:set().
dfs([], RL, RR, Open, Path, Acc) ->
    case {RL, RR, Open} of
        {0,0,0} ->
            Bin = unicode:characters_to_binary(lists:reverse(Path)),
            sets:add_element(Bin, Acc);
        _ ->
            Acc
    end;
dfs([C | Rest], RL, RR, Open, Path, Acc) ->
    case C of
        $\( ->
            Acc1 = case RL > 0 of
                       true -> dfs(Rest, RL-1, RR, Open, Path, Acc);
                       false -> Acc
                   end,
            dfs(Rest, RL, RR, Open+1, [C | Path], Acc1);
        $) ->
            Acc1 = case RR > 0 of
                       true -> dfs(Rest, RL, RR-1, Open, Path, Acc);
                       false -> Acc
                   end,
            case Open > 0 of
                true -> dfs(Rest, RL, RR, Open-1, [C | Path], Acc1);
                false -> Acc1
            end;
        _Other ->
            dfs(Rest, RL, RR, Open, [C | Path], Acc)
    end.

File: sol/solution/0301-0400/0301/Solution.dart
class Solution {
  List<String> removeInvalidParentheses(String s) {
    int rmL=0, rmR=0;
    for(int i=0;i<s.length;i++){
      if(s[i]=='(') rmL++;
      else if(s[i]==')'){ if(rmL>0) rmL--; else rmR++; }
    }
    var ans = <String>{};
    void dfs(int i, int rl, int rr, int open, String path){
      if(i==s.length){
        if(rl==0 && rr==0 && open==0) ans.add(path);
        return;
      }
      var c = s[i];
      if(c=='('){
        if(rl>0) dfs(i+1, rl-1, rr, open, path);
        dfs(i+1, rl, rr, open+1, path + c);
      } else if(c==')'){
        if(rr>0) dfs(i+1, rl, rr-1, open, path);
        if(open>0) dfs(i+1, rl, rr, open-1, path + c);
      } else {
        dfs(i+1, rl, rr, open, path + c);
      }
    }
    dfs(0, rmL, rmR, 0, "");
    var res = ans.toList();
    res.sort((a, b) => a.compareTo(b));
    return res;
  }
}
